File: Models/ConfigStore.vb
Imports System.IO
Imports System.Text
Imports System.Text.Json

''' <summary>config.json の読み書き</summary>
Public NotInheritable Class ConfigStore
    Private Sub New()
    End Sub

    Public Shared ReadOnly Property ConfigPath As String
        Get
            Return Path.Combine(AppContext.BaseDirectory, "config.json")
        End Get
    End Property

    Private Shared ReadOnly JsonOptions As New JsonSerializerOptions With {
        .WriteIndented = True,
        .Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping
    }

    Public Shared Function Load() As List(Of MacroItem)
        Dim path = ConfigPath
        If Not File.Exists(path) Then
            Dim empty As New List(Of MacroItem)()
            Save(empty)
            Return empty
        End If

        Dim json = File.ReadAllText(path, Encoding.UTF8)
        If String.IsNullOrWhiteSpace(json) Then
            Return New List(Of MacroItem)()
        End If

        Dim list = JsonSerializer.Deserialize(Of List(Of MacroItem))(json, JsonOptions)
        If list Is Nothing Then
            Return New List(Of MacroItem)()
        End If
        Return list
    End Function

    Public Shared Sub Save(macros As List(Of MacroItem))
        Dim json = JsonSerializer.Serialize(macros, JsonOptions)
        File.WriteAllText(ConfigPath, json, New UTF8Encoding(encoderShouldEmitUTF8Identifier:=False))
    End Sub

    Public Shared Function FindById(macros As IEnumerable(Of MacroItem), id As Integer) As MacroItem
        Return macros.FirstOrDefault(Function(m) m.Id = id)
    End Function

    Public Shared Function FindByName(macros As IEnumerable(Of MacroItem), name As String) As MacroItem
        Return macros.FirstOrDefault(Function(m) String.Equals(m.Name, name, StringComparison.OrdinalIgnoreCase))
    End Function

    Public Shared Function NextId(macros As IEnumerable(Of MacroItem)) As Integer
        If Not macros.Any() Then Return 1
        Return macros.Max(Function(m) m.Id) + 1
    End Function
End Class

File: Services/ConfigStore.cs
using System.Text;
using System.Text.Json;
using KeyAutomator.Models;

namespace KeyAutomator.Services;

public static class ConfigStore
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping
    };

    public static string ConfigPath => AppPaths.ConfigPath;

    public static List<MacroItem> Load()
    {
        if (!File.Exists(ConfigPath))
        {
            // 初回: 空ではなくサンプルを入れて「壊れてる？」を防ぐ
            var samples = LoadSampleMacros();
            Save(samples);
            return samples;
        }

        var json = File.ReadAllText(ConfigPath, Encoding.UTF8);
        if (string.IsNullOrWhiteSpace(json))
            return [];

        return JsonSerializer.Deserialize<List<MacroItem>>(json, JsonOptions)
               ?? throw new InvalidDataException(Loc.Get("Ex_ConfigUnreadable"));
    }

    /// <summary>
    /// exe 横 / データフォルダの <c>config.sample.json</c> または <c>config.sample.en.json</c>。
    /// 無ければ内蔵サンプル。表示名は UI 言語で上書きする。
    /// </summary>
    public static List<MacroItem> LoadSampleMacros()
    {
        foreach (var path in SampleCandidatePaths())
        {
            if (!File.Exists(path))
                continue;

            try
            {
                var json = File.ReadAllText(path, Encoding.UTF8);
                var list = JsonSerializer.Deserialize<List<MacroItem>>(json, JsonOptions);
                if (list is { Count: > 0 })
                {
                    BuiltInSamples.ApplyCurrentLanguage(list);
                    return list;
                }
            }
            catch
            {
                // 次の候補へ
            }
        }

        return BuiltInSamples.Create();
    }

    private static IEnumerable<string> SampleCandidatePaths()
    {
        var names = Loc.CurrentLanguage == Loc.English
            ? new[] { "config.sample.en.json", "config.sample.json" }
            : new[] { "config.sample.json", "config.sample.en.json" };

        foreach (var name in names)
        {
            yield return Path.Combine(AppPaths.ExeDirectory, name);
            yield return Path.Combine(AppPaths.DataDirectory, name);
        }
    }

    public static void Save(IEnumerable<MacroItem> macros)
    {
        var list = macros.ToList();
        var json = JsonSerializer.Serialize(list, JsonOptions);
        AtomicFile.WriteAllText(ConfigPath, json);

        // 書き込み後に読み返して失敗を検知する
        if (!File.Exists(ConfigPath))
            throw new IOException(Loc.Format("Ex_ConfigNotCreated", ConfigPath));

        var written = File.ReadAllText(ConfigPath, Encoding.UTF8);
        if (string.IsNullOrWhiteSpace(written))
            throw new IOException(Loc.Format("Ex_ConfigEmptyAfterSave", ConfigPath));
    }

    public static MacroItem? FindById(IEnumerable<MacroItem> macros, int id) =>
        macros.FirstOrDefault(m => m.Id == id);

    public static MacroItem? FindByName(IEnumerable<MacroItem> macros, string name) =>
        macros.FirstOrDefault(m => string.Equals(m.Name, name, StringComparison.OrdinalIgnoreCase));

    public static MacroItem? FindByAlias(IEnumerable<MacroItem> macros, string alias) =>
        macros.FirstOrDefault(m =>
            !string.IsNullOrWhiteSpace(m.Alias) &&
            string.Equals(m.Alias, alias, StringComparison.OrdinalIgnoreCase));

    public static int NextId(IEnumerable<MacroItem> macros) =>
        macros.Any() ? macros.Max(m => m.Id) + 1 : 1;
}
